src/java.base/share/classes/java/io/Reader.java

```diff
@@ -140,6 +140,125 @@ public void close() {
         };
     }
 
+    /**
+     * Returns a {@code Reader} that reads characters from a
+     * {@code CharSequence}, starting at the first character in the sequence.
+     *
+     * <p> The resulting reader is not safe for use by multiple
+     * concurrent threads. If the reader is to be used by more than one
+     * thread it should be controlled by appropriate synchronization.
+     *
+     * <p> The returned reader is initially open. The reader is closed by
+     * calling the {@code close()} method. Subsequent calls to {@code close()}
+     * have no effect.
+     *
+     * <p> After the reader has been closed, the {@code read()},
+     * {@code read(char[])}, {@code read(char[], int, int)},
+     * {@code read(CharBuffer)}, {@code ready()}, {@code skip(long)}, and
+     * {@code transferTo()} methods all throw {@code IOException}.
+     *
+     * <p> The returned reader supports the {@link #mark mark()} operation.
+     *
+     * @param cs {@code CharSequence} providing the character stream.
+     *
+     * @return a {@code Reader} which reads characters from {@code cs}
+     *
+     * @throws NullPointerException if {@code cs} is {@code null}
+     *
+     * @since 24
+     */
+    public static Reader of(final CharSequence cs) {
+        Objects.requireNonNull(cs);
+
+        return new Reader() {
+            private final int length = cs.length();
```

Comment:
As the anonymous class MUST NOT be used with be used with multiple threads, I always have seen the `CharSequence` as *fixed/static* text in the moment the `Reader` is getting used. But indeed, technically one could interleave `Reader::read()` invocations by `CharSequence.append()` (or even worse, `CharSequence.delete()`) invocations. The question is: Would that make *any* sense in the end? I mean, what happens if one has `read()` text that in the next step gets `delete()`'d? I cannot image *any* scenario where such a program would result in *useful* outcome.

&lt;fun>The fact that nobody so far (before you) brought up this question seems to proof that nobody (besides you) would write such a program. 😄 &lt;/fun>

So I would plea for clearly saying in the JavaDocs that `cs` MUST NOT be modified before `close()` is called. Every other solution implies strange side effects and slower and error-prone implementation of both, anoynous reader *and* test.

@AlanBateman WDYT?

Reply:
I would treat this specific scenario as one of the "no concurrent usage" examples. Note that by this principle, mutable objects like `StringBuilder` should not override object comparison methods as these states can change, but they do :(

Reply:
Consider the example of `StringBuffer`, which is a `CharSequence`. Wouldn't something like the following be a logical use of `Reader.of(CharSequence)`, where you create a `Reader` for the underlying sequence (which is not yet populated) and then keep reading through the `Reader` until the underlying sequence's data is finished?

```java
final StringBuffer sb = new StringBuffer();
try (final Reader reader = Reader.of(sb)) {
    final ContentGenerator cg = new ContentGenerator(sb);
    Thread.ofPlatform().start(cg);
    int numRead = 0;
    while (numRead != -1) {
        final char[] content = new char[1024];
        numRead = reader.read(content); // wait for content
    }
}

private class ContentGenerator {
    final StringBuffer sb;
    ContentGenerator(StringBuffer sb) {
        this.sb = sb;
    }

    @Override
    run() {
        while (true) {
            String foo = getSomeContent(); // maybe blocking
            sb.append(foo);
        }
    }
}
```

Reply:
>  I cannot image any scenario where such a program would result in useful outcome.

An additional point of reference is the current default implementation of `public default IntStream chars()`, which returns an IntStream of char values from the sequence and the implementation of which doesn't consider the `length()` to be fixed.

Reply:
For `chars()` or `codePoints()`, I believe calling `length()` or not was a matter of implementation convenience instead of the assumption that `length()` can change during calls. Note implementation methods in the anonymous class in `codePoints()` cache the length in local variables. Maybe they just don't want the extra field overhead in the objects they construct.

Reply:
> @AlanBateman WDYT?

It's good question as a CharSequence's length can change over time, e.g. StringBuilder. This scenario comes up regularly with InputStreams and Readers connected to files as the file may be growing and shrinking as bytes are read. It would not be surprising to read to EOF/-1 and to not read any further chars, even if the underlying sequence has grow.

Reply:
The sentence "If the sequence is modified .." in the these methods is in the context of the methods. For ofReader(CharSequence) then it is any modification, doesn't have to be concurrent. Maybe something like "If the sequence changes while the reader is open, e.g. the length changes, the behavior is undefined".

Reply:
Thank you, Alan. I have copied your wording.